pynbody/halo/subfindhdf.py

```diff
@@ -310,4 +314,186 @@ def __getattr__(self, name):
+        v=True prints out 'progress' in terms of total number of groups.
+        """
+        if self._subs is True:
+            name = 'subgrp'
+        self.base[name] = self.get_group_array()
+
+    def get_group_array(self):
+        ar = np.zeros(len(self.base), dtype=int) - 1
+        if self._subs is True:
+            offs = self._subhalodat['SubhaloOffsetType'][:, self.ptypenum]
+            ls = self._subhalodat['SubhaloLenType'][:, self.ptypenum]
+        else:
+            offs = self._halodat['GroupOffsetType'][:, self.ptypenum]
+            ls = self._halodat['GroupLenType'][:, self.ptypenum]
+
+        for i in range(len(offs)):
+            ar[offs[i]:offs[i] + ls[i]] = i
+        return ar
+
+    def extract_prop_w_units(self, simarray, elem):
+        if type(simarray) == array.SimArray:
+            if len(simarray.shape) > 1:
+                return simarray[elem]
+            else:
+                simarray_i = array.SimArray([simarray[elem]], simarray.units)
+                simarray_i.sim = simarray.sim
+                return simarray_i
+        else:
+            return simarray[elem]
+
+    def get_halo_properties(self, i):
+        properties = {}
+        if self._subs is False:
+            for key in self._keys:
+                properties[key] = self.extract_prop_w_units(self._halodat[key], i)
+                properties['children'] = np.where(self._subhalodat['SubhaloGroupNr'] == i)[0]
+        else:
+            for key in self._keys:
+                properties[key] = self.extract_prop_w_units(self._subhalodat[key], i)
+        return properties
+
+    def _get_halo(self,i):
+        if self._subs is True:
+            offset = self._subhalodat['SubhaloOffsetType'][i, self.ptypenum]
+            length = self._subhalodat['SubhaloLenType'][i, self.ptypenum]
+        else:
+            offset = self._halodat['GroupOffsetType'][i, self.ptypenum]
+            length = self._halodat['GroupLenType'][i, self.ptypenum]
+
+        x = Halo(i, self, self.base, np.where(np.in1d(self.base['iord'], self.ids[offset: offset + length]))[0])
+        x._descriptor = "halo_" + str(i)
+        x.properties.update(self.get_halo_properties(i))
+        return x
+
+    def _readheader(self):
+        """ Load the group catalog header. """
+        with h5py.File(self.halofilename, 'r') as f:
+            header = dict(f['Header'].attrs.items())
+        return header
+
+    def _readparams(self):
+        """ Load the group catalog header. """
+        with h5py.File(self.halofilename, 'r') as f:
+            header = dict(f['Parameters'].attrs.items())
+        return header
+
+    def _read_ids(self):
+        data_ids = np.array([], dtype=self.dtype_int)
+        for n in range(self._tasks):
+            ids = self.base._hdf_files[n][self.ptype]['ParticleIDs']
+            data_ids = np.append(data_ids, ids)
+        return data_ids
+
+    def _read_groups(self):
```

Comment:
Could you add a couple of comments or split up into a couple of methods to clarify what is happening here? (I realise other loaders are not great at this but we might as well improve the code standard.)